File: typescript/src/patterns/strategy/ConcreteStrategies.ts
import type { PriceStrategy } from './Strategy.js';

export class PercentageDiscount implements PriceStrategy {
  constructor(private percent: number) {
    if (percent < 0 || percent > 100) throw new Error('percent must be 0..100');
  }
  apply(price: number): number {
    return +(price * (1 - this.percent / 100)).toFixed(2);
  }
}

export class FixedDiscount implements PriceStrategy {
  constructor(private amount: number) {}
  apply(price: number): number {
    return Math.max(0, +(price - this.amount).toFixed(2));
  }
}

File: python/src/patterns/strategy/concrete_strategies.py
from .strategy import PriceStrategy

class PercentageDiscount(PriceStrategy):
    def __init__(self, percent: float):
        if percent < 0 or percent > 100:
            raise ValueError("percent must be 0..100")
        self.percent = percent
    def apply(self, price: float) -> float:
        return round(price * (1 - self.percent / 100), 2)

class FixedDiscount(PriceStrategy):
    def __init__(self, amount: float):
        self.amount = amount
    def apply(self, price: float) -> float:
        return max(0.0, round(price - self.amount, 2))
